### PR TITLE
Issue #376  Define CRUD operation for feature flag

### PR DESCRIPTION
**Pull request by Ankita2002-Fr**
@Cartmanishere  @ankushdharkar @ASHWINIAJIT

I have worked on the CRUD operations for the feature flag and also added authentication and validation for the same.

Please review the PR!

### CODE REVIEW

File: controllers/featureFlags.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+
+const addFeatureFlag = async (req, res) => {
+  try {
+    const featureFlag = await featureFlagQuery.addFeatureFlags(req.body, req.userData.username)
+    return res.json({
+      message: 'FeatureFlag added successfully',
+      featureFlag
```
Comment (Cartmanishere) [inline controllers/featureFlags.js:35]: This should be passed under `data`.

```
{
  message: "..."
  data: <featureFlag>
}
```

File: middlewares/validators/featureFlags.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  } catch (error) {
+    logger.error(`Error in validating featureFlag data: ${error}`)
+    res.boom.badRequest(error.details[0].message)
+  }
+}
+
+const updateFeatureFlags = async (req, res, next) => {
+  const schema = joi.object().keys({
+    title: joi.string().optional(),
+    config: joi.object().required()
```
Comment (Cartmanishere) [inline middlewares/validators/featureFlags.js:25]: Add validation for `enabled` field inside `config` object.

You can search for how to validate nested keys in joi.

File: models/featureFlags.js [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+      })
+    })
+    const users = []
+    const result = {}
+
+    featureFlags.forEach((item) => {
+      if (!users.includes(item.owner)) { users.push(item.owner) }
+    })
+
+    const getImage = async (usersData) => {
```
Comment (Cartmanishere) [inline models/featureFlags.js:27]: this function is not needed. You can directly call `fetchUserImage` in the loop below.

File: models/users.js [hunk 4]
```diff
@@ -167,11 +167,27 @@ const updateUserPicture = async (image, userId) => {
...
+ *
+ * @param users {array}
+ */
+const fetchUserImage = async (users) => {
+  const data = await userModel.where('username', 'in', users).get()
+  const hash = {}
+  data.forEach((item) => {
+    hash[item.data().username] = item.data().img
+  })
+  // console.log(hash)
```
Comment (Cartmanishere) [inline models/users.js:181]: Remove the console.

File: models/featureFlags.js [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+      return await userModel.fetchUserImage(usersData)
+    }
+
+    let start = 0
+    let end = 10
+
+    for (let i = 0; i < Math.ceil(users.length / 10); i++) {
+      const usersData = users.slice(start, end)
```
Comment (Cartmanishere) [inline models/featureFlags.js:35]: You can also do the logic in loop using just `i`.

Don't need `start` and `end` vars.

**Review by Cartmanishere — CHANGES_REQUESTED** [review]
Add tests for these as well.

File: controllers/featureFlags.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+ * Fetches all the featureFlag
+ *
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+
+const getFeatureFlags = async (req, res) => {
+  try {
+    const allFeatureFlags = await featureFlagQuery.fetchFeatureFlag()
+    // console.log(res.status);
```
Comment (swarajpure) [inline controllers/featureFlags.js:13]: Commented code...?

Comment (Ankita2002-Fr) [reply controllers/featureFlags.js:13]: Removed!

**Review by Cartmanishere — DISMISSED** [review]

**Comment by Cartmanishere** [comment]
@Ankita2002-Fr Looks like your tests are failing due to unrelated change. See if you fix the test as well.

Also please, resolve the conflict merge conflict and update the PR.

Rest lgtm. We can merge it.

**Review by Cartmanishere — APPROVED** [review]

**Pull request merged**